### PR TITLE
use MIX_HOME,HEX_HOME when installing mix/hex dependencies

### LINKED ISSUE #194: Buildpack incompatible with building in /app
Hi

I'm on the team that maintains Heroku's build system and official buildpacks, and wanted to let you know about a future incompatibility with this buildpack.

The directory in which the Heroku build system performs builds is currently a path like`/tmp/build_<hash>`.

In the near future this path will be changing to `/app` so that the build-time and run-time app locations are the same path - in order to resolve a number of long standing bugs and reduce the number of hacks buildpacks have to use to work around non-relocatable languages/toolchains. As part of this change, `HOME` will also be changing from `/app` to a PATH under `/tmp` (otherwise temporary files/secrets saved to HOME would end up in the slug).

Trying out the above change with this buildpack (using [this sample app](https://github.com/HashNuke/heroku-buildpack-elixir-test) and `heroku labs:enable build-in-app-dir`) revealed that builds with this buildpack will break - or more accurately, output an error to the build log, however the build be incorrectly marked as successful and instead the app give a boot time error.

Build time error:

```
remote: -----> Building on the Heroku-20 stack
remote: -----> Performing build in /app since 'build-in-app-dir' is enabled
remote: -----> Elixir app detected
remote: -----> Will export the following config vars:
remote:        * MIX_ENV=prod
remote: -----> Checking Erlang and Elixir versions
remote:        WARNING: elixir_buildpack.config wasn't found in the app
remote:        Using default config from Elixir buildpack
remote:
remote:        IMPORTANT: The default elixir_version will be removed on 2021-06-01. Please explicitly set an elixir_version in your elixir_buildpack.config before then or your deploys will fail.
remote:
remote:        Will use the following versions:
remote:        * Stack heroku-20
remote:        * Erlang 20.1
remote:        * Elixir v1.5.3
remote: -----> Using cached Erlang 20.1
remote: -----> Installing Erlang 20.1
remote:
remote: cp: will not create hard link '/app/.platform_tools/erlang/erlang/erlang' to directory '/app/.platform_tools/erlang/erlang'
remote: -----> Using cached Elixir v1.5.3
...
```

Then at boot:

```
2021-03-14T18:56:25.000000+00:00 app[api]: Build succeeded
2021-03-14T18:56:26.039686+00:00 app[web.1]: /app/.platform_tools/elixir/bin/elixir: 127: exec: erl: not found
2021-03-14T18:56:26.117294+00:00 heroku[web.1]: Process exited with status 127
2021-03-14T18:56:26.191815+00:00 heroku[web.1]: State changed from starting to crashed
```

It appears there are two issues:
1. Errors during the build don't cause the build to fail (not using Bash exit on error)
2. The file removals, hardlink and copy here need to take into account that the BUILD_DIR might be `/app`: https://github.com/HashNuke/heroku-buildpack-elixir/blob/b9092d471ebb0163dbf1bf733b719cf1e31c8f64/lib/erlang_funcs.sh#L29-L38

We're hoping to enable the new build behaviour very soon - and whilst there will be an opt-out labs, it would be great to get this resolved before then to prevent Erlang apps from failing to boot after the first build, for users that miss the Heroku changelog entry etc.

Many thanks :-)

### PR DESCRIPTION
**Pull request by fivetanley**
tl;dr:

- Instead of relying on Mix's default behavior of installing to `$HOME/.mix`, the buildpack sets `$MIX_HOME` to a directory controlled by the buildpack when running `mix`
- Instead of relying on Hex's default behavior of installing to `$HOME/.hex`, the buildpack sets `$HEX_HOME` to a directory controlled by the buildpack when running `hex`
- The buildpack now exports `HEX_HOME` and `MIX_HOME` if they are not set as configuration variables.

By using `MIX_HOME` and `HEX_HOME`, we can consolidate some logic for transferring mix/hex files in and out of the cache and app build directory.

This should allow the buildpack to work in the new build system and the old build system since this buildpack no longer relies on what the value of `HOME` is at either build time or runtime.

fixes #194

https://github.com/HashNuke/heroku-buildpack-elixir/issues/194

# The Longer Explanation

This strategy was taken from the [heroku-buildpack-ruby buildpack][1], which sets [GEM_HOME and GEM_PATH environment variables][2] as well as the [`BUNDLE_PATH` environment variable][3] to a path in the `BUILD_DIR` argument when running the `bundle install` command.

Although the Elixir programming langauge and buildpack doesn't use Bundler/Rubygems, it has similar patterns for setting the installation directory and archive paths for Mix and Hex.

`MIX_HOME` has been supported since `v0.7.0` of Elixir ([support for `MIX_HOME` added ~9 years ago][4]).

Note: There is a `MIX_ARCHIVES` env variable that Mix can use, but [MIX_ARCHIVES support was added in version 0.14.7][5]. I'm not sure if anyone is still using this Elixir version (they should upgrade as the OTP that 0.14.7 supports is very outdated and doesn't receive security updates). If `MIX_ARCHIVES` is not present, then [Mix will put archives in `$MIX_HOME/archives`][6]. In order to keep compatibility as close as possible, this commit continues using the old strategy of backing up everything in `MIX_HOME`, rather than what's in `MIX_ARCHIVES` alone.

[`HEX_HOME` support was added in Hex v0.3.3][7]. This was the commit that made Hex put stuff in `$HEX_HOME` directory, meaning that support for `HEX_HOME` has existed for a long time and we should be able to rely on Hex using the `HEX_HOME` directory. The old strategy used by the Elixir buildpack always backed up the default `HEX_HOME` value (`$HOME/.hex`).

[1]: https://github.com/heroku/heroku-buildpack-ruby
[2]: https://github.com/heroku/heroku-buildpack-ruby/blob/52e02d56370c1e05434bc550a06f1bcde5849644/lib/language_pack/ruby.rb#L324-L325
[3]: https://github.com/heroku/heroku-buildpack-ruby/blob/52e02d56370c1e05434bc550a06f1bcde5849644/lib/language_pack/ruby.rb#L349
[4]: https://github.com/elixir-lang/elixir/commit/e518c5263e66d6554c7229d170a94b8afac25f70#diff-ff4ab450307d771fedb0471df59d9bd71f4f90559375d57f8015fda8ba4cf301R21
[5]: https://github.com/elixir-lang/elixir/commit/02a415facdfc86373eebaf507b83061b291fd6d2#diff-75787cbe3255513265f665bb859b99a624db2b56bec01e48455cd63577ad8cadR16
[6]: https://github.com/elixir-lang/elixir/blob/c57d08a32f1308769a2fbaf7a2fe385ab2fb5d22/lib/mix/lib/mix.ex#L463
[7]: https://github.com/hexpm/hex/commit/5dd8ae020dc4c31bc662e490c79221423fc5d9f6

### CODE REVIEW

**Comment by fivetanley** [comment]
@edmorley I've tested this on Heroku CI, Heroku-20 (old build system), and Heroku-20 (new build system). Can you take a peak and let me know what you think? I think i addressed your [feedback from the reverted PR](https://github.com/HashNuke/heroku-buildpack-elixir/pull/200#discussion_r649766001), but let me know if I missed something or misunderstood.

File: lib/path_funcs.sh [hunk 1]
```diff
@@ -2,40 +2,40 @@ function platform_tools_path() {
   echo "${build_path}/.platform_tools"
 }
 
-function erlang_path() {
```
Comment (fivetanley) [inline lib/path_funcs.sh:5]: i added prefixes to the names of these functions to make it easier to determine what's "build-time" (when `bin/compile` runs) and what's "runtime" (values present at runtime when the app actually runs).

File: lib/path_funcs.sh [hunk 2]
```diff
@@ -2,40 +2,40 @@ function platform_tools_path() {
...
+
+function runtime_hex_home_path() {
+  echo "${runtime_path}/.hex"
+}
+
+function build_mix_home_path() {
+  echo "${build_path}/.mix"
+}
 
-function erlang_build_path() {
```
Comment (fivetanley) [inline lib/path_funcs.sh:29]: since this was only used in one place, i moved it to a `local` variable in `install_erlang` to remove confusion and having multiple functions named after the erlang build path.

File: lib/app_funcs.sh [hunk 3]
```diff
@@ -14,27 +14,17 @@ function restore_app() {
 
 
 function copy_hex() {
-  mkdir -p ${build_path}/.mix/archives
-  mkdir -p ${build_path}/.hex
-
-
-  # hex is a directory from elixir-1.3.0
-  full_hex_file_path=$(ls -dt ${HOME}/.mix/archives/hex-* | head -n 1)
```
Comment (fivetanley) [inline lib/app_funcs.sh:22]: Finding out the full_hex_file_path isn't necessary; setting `MIX_HOME` and `HEX_HOME` means it will end up in the cache and in the build slug since Mix will install hex using `MIX_HOME`.

**Comment by fivetanley** [comment]
@HashNuke @jesseshieh any thoughts on this approach?

**Comment by jesseshieh** [comment]
At first glance, it looks great to me. Sorry, I was waiting for @edmorley to comment first. As long as it's tested to work on both the old system and the new system, and @edmorley is happy with it, then I'm happy with it.

File: lib/app_funcs.sh [hunk 4]
```diff
@@ -148,33 +139,47 @@ function pre_compile_hook() {
...
-  # Only write MIX_ENV to export if the application did not set MIX_ENV
-  if [ ! -f $env_path/MIX_ENV ]; then
-    export_line="${export_line}
-                 export MIX_ENV=${MIX_ENV}"
-  fi
+function write_export() {
+  output_section "Writing export for multi-buildpack support"
 
-  echo $export_line > $build_pack_path/export
+  echo_profile_env_vars >> "${build_pack_path}/export"
```
Comment (edmorley) [inline lib/app_funcs.sh:184]: If I'm following this correctly, it seems like the export file will contain runtime paths in `MIX_HOME` and `HEX_HOME`, rather than build time paths. When the export file is sourced prior to later buildpacks running, it's still build time, so the slug won't yet be extracted to `/app`. This might work fine since I see there are also copying steps that copy files to `/app` during the build, so `/app` may still contain what you need when later buildpacks run (`/app` is not purged between buildpacks) - but wanted to mention in case.

**Review by edmorley — COMMENTED** [review]
Sorry I missed the @ mention.

This looks good to me! (Just the one comment but may be fine as is)
Thank you for working on this! :-)

**Comment by jesseshieh** [comment]
I just tried testing this and although this buildpack still works on it's own, when paired with the [phoenix buildpack](https://github.com/gjaldon/heroku-buildpack-phoenix-static), I get this error.
```
       Could not find Hex, which is needed to build dependency :phoenix
       Shall I install Hex? (if running non-interactively, use "mix local.hex --force") [Yn] ** (Mix) Could not find an SCM for dependency :phoenix from Foo.MixProject
```

**Comment by edmorley** [comment]
Yeah that's likely due to:
https://github.com/HashNuke/heroku-buildpack-elixir/pull/203#discussion_r662209951

File: lib/app_funcs.sh (hunk 4 above)
Comment (fivetanley) [reply lib/app_funcs.sh:184]: Good catch; I've updated the mix install and erlang install functions to copy if the build time path and the runtime path aren't the same.

**Comment by fivetanley** [comment]
@jesseshieh I've updated this PR to copy the Hex/Mix/Elixir directories as well as the erlang directory to the runtime path if the runtime path and buildtime path don't match. I tried this with an application with the phoenix buildpack (thanks for that detail! our app doesn't happen to use the phoenix buildpack since we don't use static assets) and it seems to work. Can you please give this PR another try when you have a moment? Thanks!

File: lib/app_funcs.sh (hunk 4 above)
Comment (edmorley) [reply lib/app_funcs.sh:184]: @fivetanley What would you say are the pros/cons of the copying approach vs making sure the `export` file has the correct build-time `MIX_HOME` and `HEX_HOME` values? (Currently the `export` file has the run-time `/app` values).

With the copying approach, if a later buildpack eg installs additional dependencies, won't those end up in the `/app` copy, and not in build_dir, and so be missing at runtime?

Comment (edmorley) [reply lib/app_funcs.sh:184]: > vs making sure the `export` file has the correct build-time `MIX_HOME` and `HEX_HOME` values

To be clear, by this approach I mean copying `echo_profile_env_vars` and having another version that outputs paths that use `$BUILD_DIR` for use with the `export` file, rather than trying to use the output of `echo_profile_env_vars` for both the `profile.d` script and the `export` file.

Comment (fivetanley) [reply lib/app_funcs.sh]: Ah, ok, I think I understand now. Can you review the updated diff (this convo is on the outdated diff) for this function?

Comment (edmorley) [reply lib/app_funcs.sh]: @fivetanley The latest version looks great - thank you! :-)

**Comment by jesseshieh** [comment]
Looks great and thank you so much for doing this!
